pkg/gcp/actuator/policy.go

```diff
@@ -0,0 +1,248 @@
+// ensurePolicyBindings will add and remove any policy bindings for the service account to match the
+// roles list provided.
+func ensurePolicyBindings(rootClient ccgcp.Client, roles []string, svcAcct *iamadminpb.ServiceAccount) error {
+	needPolicyUpdate := false
+
+	projectName := rootClient.GetProjectName()
+	policy, err := rootClient.GetProjectIamPolicy(projectName, &cloudresourcemanager.GetIamPolicyRequest{})
+
+	if err != nil {
+		return fmt.Errorf("error fetching policy for project: %v", err)
+	}
+
+	// Validate that each role exists, and add the policy binding as needed
+	for _, definedRole := range roles {
+		// Check it exists
+		role, err := rootClient.GetRole(context.TODO(), &iamadminpb.GetRoleRequest{
+			Name: definedRole,
+		})
+		if err != nil {
+			if status.Code(err) == codes.NotFound {
+				return fmt.Errorf("role %s not found: %v", role, err)
+			} else {
+				return fmt.Errorf("error checking if role exists: %v", err)
+			}
+		}
+
+		// Add policy binding
+		modified := addPolicyBinding(policy, role, svcAcct)
+		if modified {
+			needPolicyUpdate = true
+		}
+
+	}
+
+	// Remove extra role bindings as needed
+	modified := purgeExtraPolicyBindings(policy, roles, svcAcct)
+	if modified {
+		needPolicyUpdate = true
+	}
+
+	if needPolicyUpdate {
+		if rootClient == nil {
+			return fmt.Errorf("detected need for policy update, but no root creds available")
+		}
+		return setProjectIamPolicy(rootClient, policy, projectName)
+	}
+
+	// If we made it this far there were no updates needed
+	return nil
+}
+
+func purgeExtraPolicyBindings(policy *cloudresourcemanager.Policy, roleList []string, svcAcct *iamadminpb.ServiceAccount) bool {
+	modifiedPolicy := false
+	svcAcctBindingName := serviceAccountBindingName(svcAcct)
+
+	for _, binding := range policy.Bindings {
+		// find if our service account has an entry in this binding
+		for j, member := range binding.Members {
+			if member == svcAcctBindingName {
+				removeMember := true
+
+				// check if this role is one that should be bound to this service account's roleList
+				for _, role := range roleList {
+					if role == binding.Role {
+						removeMember = false
+						break
+					}
+				}
+
+				if removeMember {
+					// It is okay to submit a policy with a binding entry where the member list
+					// is empty. The policy will be cleaned up on the GCP-side and it will be
+					// as if we had removed the entire binding entry.
+
+					binding.Members = append(binding.Members[:j], binding.Members[j+1:]...)
+					modifiedPolicy = true
+				}
+			}
+		}
+	}
+
+	return modifiedPolicy
+}
+
+func addPolicyBinding(policy *cloudresourcemanager.Policy, role *iamadminpb.Role, svcAcct *iamadminpb.ServiceAccount) bool {
+	for i, binding := range policy.Bindings {
+		if binding.Role == role.Name {
+			return addServiceAccountToBinding(svcAcct, policy.Bindings[i])
+		}
+	}
+
+	// if we didn't find an existing binding entry, then make one
+	createServiceAccountRoleBinding(policy, role, svcAcct)
```

Comment:
What is the difference between role binding and policy binding?

Reply:
A policy contains a list of bindings, and each binding entry is what ties a specific role to a list of users/members:
Here is a list of binding entries for everything related to AppEngine:

```
[jdiaz@minigoomba ~]$ gcloud projects get-iam-policy openshift-hive-dev --format=json | jq  '.bindings[] | select(.role | contains("roles/appengine"))'
{
  "members": [
    "serviceAccount:jdiaz-tect-6-my-test-cre-4jl2w@openshift-hive-dev.iam.gserviceaccount.com",
    "serviceAccount:test-serviceaccount-12345@openshift-hive-dev.iam.gserviceaccount.com"
  ],
  "role": "roles/appengine.appAdmin"
}
{
  "members": [
    "serviceAccount:jdiaz-tect-6-my-test-cre-4jl2w@openshift-hive-dev.iam.gserviceaccount.com"
  ],
  "role": "roles/appengine.appViewer"
}
```
Two binding entries above. You can see that the appAdmin role is bound to two service accounts, and the appViewer role is bound to one service account.

If I've used the term 'policy binding' somewhere, that's probably a mistake.